File: mobile/app/year-end-tax.js
// Year-end tax check for employees: am I getting a refund, or do I still owe?
// Every December an employer trues up the whole year's tax, and steady monthly
// withholding rarely matches the real annual tax once a 13th month, bonuses, a
// raise, or a mid-year start are counted. This estimates the annual tax due and
// compares it to what was already withheld. All math is in lib/phtax.js (pure
// and tested). This is an estimate, not your Form 2316, and the screen says so.

import { useMemo, useState } from 'react';
import { Pressable, ScrollView, StyleSheet, Text, TextInput, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useRouter } from 'expo-router';
import { Ionicons } from '@expo/vector-icons';
import { spacing, radius, fontSize, fontWeight } from '../theme';
import { useTheme } from '../context/Theme';
import { formatMoney } from '../lib/format';
import { annualizeCompensation, BONUS_TAX_FREE_CEILING, RATES_YEAR } from '../lib/phtax';

export default function YearEndTax() {
  const { colors } = useTheme();
  const styles = useMemo(() => makeStyles(colors), [colors]);
  const router = useRouter();

  const [basic, setBasic] = useState('');
  const [allowance, setAllowance] = useState('');
  const [monthsWorked, setMonthsWorked] = useState('');
  const [bonuses, setBonuses] = useState('');
  const [withheld, setWithheld] = useState('');

  const parse = (s) => Number(String(s).replace(/[, ]/g, '')) || 0;
  const basicNum = parse(basic);
  const allowanceNum = parse(allowance);
  const monthsNum = monthsWorked === '' ? 12 : parse(monthsWorked);
  const bonusesNum = parse(bonuses);
  const withheldNum = parse(withheld);

  const r = useMemo(
    () =>
      annualizeCompensation(basicNum, {
        taxableAllowance: allowanceNum,
        monthsWorked: monthsNum,
        bonuses: bonusesNum,
        taxWithheld: withheldNum,
      }),
    [basicNum, allowanceNum, monthsNum, bonusesNum, withheldNum]
  );
  const m = (n) => formatMoney(Math.round(n));

  const ready = basicNum > 0;
  // Round before deciding the verdict so a sub-peso gap reads as settled, not a
  // stray one peso refund or shortfall.
  const gap = Math.round(r.difference);
  const settled = gap === 0;
  const refund = gap > 0;

  return (
    <SafeAreaView style={styles.screen} edges={['top']}>
      <View style={styles.headerBar}>
        <Pressable onPress={() => router.back()} hitSlop={10} style={styles.back}>
          <Ionicons name="chevron-back" size={24} color={colors.text} />
        </Pressable>
        <Text style={styles.headerTitle}>Year-end tax check</Text>
        <View style={{ width: 24 }} />
      </View>

      <ScrollView contentContainerStyle={styles.content} keyboardShouldPersistTaps="handled">
        <Text style={styles.intro}>
          At year end your employer adds up your whole year and settles the tax. If too much was taken from your payslips you get a refund, if too little you pay the small difference. See roughly which way you are heading.
        </Text>

        <Text style={styles.fieldLabel}>Monthly basic pay</Text>
        <View style={styles.inputWrap}>
          <Text style={styles.peso}>₱</Text>
          <TextInput style={styles.input} value={basic} onChangeText={setBasic} keyboardType="numeric" placeholder="e.g. 25,000" placeholderTextColor={colors.faint} autoFocus />
        </View>

        <View style={styles.twoCol}>
          <View style={styles.col}>
            <Text style={styles.fieldLabel}>Monthly taxable allowance</Text>
            <View style={styles.smallInputWrap}>
              <Text style={styles.pesoSm}>₱</Text>
              <TextInput style={styles.smallInput} value={allowance} onChangeText={setAllowance} keyboardType="numeric" placeholder="0" placeholderTextColor={colors.faint} />
            </View>
          </View>
          <View style={styles.col}>
            <Text style={styles.fieldLabel}>Months worked this year</Text>
            <View style={styles.smallInputWrap}>
              <TextInput style={styles.smallInput} value={monthsWorked} onChangeText={setMonthsWorked} keyboardType="numeric" placeholder="12" placeholderTextColor={colors.faint} />
            </View>
          </View>
        </View>

        <Text style={[styles.fieldLabel, { marginTop: spacing.lg }]}>13th month plus bonuses this year</Text>
        <View style={styles.inputWrap}>
          <Text style={styles.peso}>₱</Text>
          <TextInput style={styles.input} value={bonuses} onChangeText={setBonuses} keyboardType="numeric" placeholder="e.g. 25,000" placeholderTextColor={colors.faint} />
        </View>
        <Text style={styles.subHint}>The first {m(BONUS_TAX_FREE_CEILING)} of your 13th month and other bonuses combined is tax free. Only the excess is taxed.</Text>

        <Text style={[styles.fieldLabel, { marginTop: spacing.lg }]}>Income tax withheld so far</Text>
        <View style={styles.inputWrap}>
          <Text style={styles.peso}>₱</Text>
          <TextInput style={styles.input} value={withheld} onChangeText={setWithheld} keyboardType="numeric" placeholder="Total for the year" placeholderTextColor={colors.faint} />
        </View>
        <Text style={styles.subHint}>Add up the tax taken from each payslip this year, or read it from your latest payslip year to date. Your Form 2316 shows the final figure.</Text>

        {ready ? (
          <>
            <View style={[styles.card, refund ? styles.cardGood : settled ? styles.cardNeutral : styles.cardOwe]}>
              <Text style={styles.amtLabel}>
                {settled ? 'You are about even' : refund ? 'Estimated refund' : 'You may still owe about'}
              </Text>
              <Text style={[styles.amtValue, refund ? styles.good : settled ? styles.neutral : styles.owe]}>
                {settled ? m(0) : m(Math.abs(r.difference))}
              </Text>
              <Text style={styles.amtNote}>
                {settled
                  ? 'What was withheld matches your estimated tax for the year, so expect little or no adjustment.'
                  : refund
                    ? 'More was withheld than your estimated tax for the year, so you should get the difference back.'
                    : 'Less was withheld than your estimated tax, so a small amount may be deducted at year end.'}
              </Text>
            </View>

            <View style={styles.breakdownCard}>
              <View style={styles.row}>
                <Text style={styles.rowLabel}>Taxable income for the year</Text>
                <Text style={styles.rowValue}>{m(r.annualTaxable)}</Text>
              </View>
              {r.bonusTaxable > 0 ? (
                <View style={styles.row}>
                  <Text style={styles.rowLabel}>Includes taxable part of bonuses</Text>
                  <Text style={styles.rowSubtle}>{m(r.bonusTaxable)}</Text>
                </View>
              ) : null}
              <View style={styles.row}>
                <Text style={styles.rowLabel}>Tax due for the year</Text>
                <Text style={styles.rowValue}>{m(r.annualTaxDue)}</Text>
              </View>
              <View style={styles.row}>
                <Text style={styles.rowLabel}>Tax already withheld</Text>
                <Text style={styles.rowValue}>{m(r.taxWithheld)}</Text>
              </View>
              <View style={styles.netRow}>
                <Text style={styles.netLabel}>{refund ? 'Refund to you' : settled ? 'Difference' : 'Still to pay'}</Text>
                <Text style={[styles.netValue, refund ? styles.good : settled ? styles.neutral : styles.owe]}>
                  {settled ? m(0) : m(Math.abs(r.difference))}
                </Text>
              </View>
              <View style={styles.rateRow}>
                <Text style={styles.rateText}>Effective tax rate about {r.effectiveRate}% of what you earned this year.</Text>
              </View>
            </View>

            <View style={styles.infoCard}>
              <Text style={styles.infoKicker}>GOOD TO KNOW</Text>
              <Text style={styles.infoLine}>
                Most employees do not file this themselves. Your employer does the year-end adjustment and a refund usually shows up in your December or January pay. This just tells you what to expect.
              </Text>
              <Text style={styles.infoLine}>
                If you changed jobs this year, give your new employer your Form 2316 from the old one so they annualize correctly, otherwise you can be under-withheld.
              </Text>
            </View>
          </>
        ) : (
          <Text style={styles.hint}>Enter your monthly basic pay to see your year-end estimate.</Text>
        )}

        <Text style={styles.disclaimer}>
          Estimate based on {RATES_YEAR} BIR rules (TRAIN graduated table, the {m(BONUS_TAX_FREE_CEILING)} tax-free ceiling for 13th month and bonuses, and mandatory contributions figured on basic pay). It assumes a steady salary and that this job is your only income. Your real refund depends on your exact payslips, allowances, and any de minimis benefits. Not a substitute for your Form 2316 or an accountant.
        </Text>
      </ScrollView>
    </SafeAreaView>
  );
}

function makeStyles(colors) {
  return StyleSheet.create({
    screen: { flex: 1, backgroundColor: colors.background },
    headerBar: { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', paddingHorizontal: spacing.lg, paddingTop: spacing.md, paddingBottom: spacing.sm },
    back: { marginLeft: -4 },
    headerTitle: { color: colors.text, fontSize: fontSize.subtitle, fontWeight: fontWeight.bold },
    content: { padding: spacing.lg, paddingBottom: spacing.xxl },
    intro: { color: colors.muted, fontSize: fontSize.small, lineHeight: 19, marginBottom: spacing.lg },

    fieldLabel: { color: colors.muted, fontSize: fontSize.caption, marginBottom: spacing.sm, letterSpacing: 0.3 },
    inputWrap: { flexDirection: 'row', alignItems: 'center', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, paddingHorizontal: spacing.lg },
    peso: { color: colors.muted, fontSize: fontSize.title, fontWeight: fontWeight.bold, marginRight: spacing.sm },
    input: { flex: 1, paddingVertical: spacing.md, color: colors.text, fontSize: fontSize.title, fontWeight: fontWeight.bold },
    subHint: { color: colors.faint, fontSize: fontSize.caption, lineHeight: 16, marginTop: spacing.sm },

    twoCol: { flexDirection: 'row', gap: spacing.md, marginTop: spacing.lg },
    col: { flex: 1 },
    smallInputWrap: { flexDirection: 'row', alignItems: 'center', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, paddingHorizontal: spacing.md },
    pesoSm: { color: colors.muted, fontSize: fontSize.body, marginRight: spacing.xs },
    smallInput: { flex: 1, paddingVertical: spacing.md, color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold },

    card: { borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.lg, marginTop: spacing.xl },
    cardGood: { backgroundColor: colors.card, borderColor: colors.primary },
    cardNeutral: { backgroundColor: colors.card, borderColor: colors.border },
    cardOwe: { backgroundColor: colors.card, borderColor: colors.border },
    amtLabel: { color: colors.muted, fontSize: fontSize.caption, letterSpacing: 0.3 },
    amtValue: { fontSize: fontSize.huge, fontWeight: fontWeight.heavy, marginTop: 2 },
    amtNote: { color: colors.textSecondary, fontSize: fontSize.small, lineHeight: 18, marginTop: spacing.sm },
    good: { color: colors.primary },
    neutral: { color: colors.text },
    owe: { color: colors.warning || colors.text },

    breakdownCard: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.lg, marginTop: spacing.md },
    row: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', paddingVertical: spacing.sm },
    rowLabel: { color: colors.muted, fontSize: fontSize.small, flexShrink: 1, paddingRight: spacing.md },
    rowValue: { color: colors.text, fontSize: fontSize.small, fontWeight: fontWeight.medium },
    rowSubtle: { color: colors.textSecondary, fontSize: fontSize.small },
    netRow: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', borderTopColor: colors.border, borderTopWidth: 1, marginTop: spacing.sm, paddingTop: spacing.md },
    netLabel: { color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold },
    netValue: { fontSize: fontSize.subtitle, fontWeight: fontWeight.heavy },
    rateRow: { borderTopColor: colors.border, borderTopWidth: StyleSheet.hairlineWidth, marginTop: spacing.sm, paddingTop: spacing.md },
    rateText: { color: colors.textSecondary, fontSize: fontSize.small, lineHeight: 18 },

    infoCard: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.lg, marginTop: spacing.md },
    infoKicker: { color: colors.muted, fontSize: fontSize.caption, fontWeight: fontWeight.bold, letterSpacing: 1.2, marginBottom: spacing.sm },
    infoLine: { color: colors.textSecondary, fontSize: fontSize.small, lineHeight: 19, marginTop: spacing.xs },

    hint: { color: colors.muted, fontSize: fontSize.body, textAlign: 'center', marginTop: spacing.xl, marginBottom: spacing.xl },
    disclaimer: { color: colors.faint, fontSize: fontSize.caption, lineHeight: 17, marginTop: spacing.xl },
  });
}

File: flutter/lib/screens/year_end_tax.dart
// Year-end tax check for employees: am I getting a refund, or do I still owe?
//
// Every December an employer trues up the whole year's tax, and steady monthly
// withholding rarely matches the real annual tax once a 13th month, a bonus, a
// raise, or a mid-year start are counted. This estimates the annual tax due
// and compares it against what was already withheld.
//
// Every peso comes from money/phtax.dart, which is golden locked to the RN
// engine. This screen collects five numbers and reads the answer back. It is
// an estimate, not a Form 2316, and it says so where it cannot be missed
// rather than in fine print at the bottom.

import 'package:flutter/material.dart';

import '../data/store.dart';
import '../money/debtmath.dart' show formatMoneyText;
import '../money/phtax.dart';
import '../theme.dart';
import '../widgets/section.dart';

class YearEndTaxScreen extends StatefulWidget {
  final SalapifyStore store;
  const YearEndTaxScreen({super.key, required this.store});

  @override
  State<YearEndTaxScreen> createState() => _YearEndTaxScreenState();
}

class _YearEndTaxScreenState extends State<YearEndTaxScreen> {
  final _basic = TextEditingController();
  final _allowance = TextEditingController();
  final _months = TextEditingController();
  final _bonuses = TextEditingController();
  final _withheld = TextEditingController();

  @override
  void dispose() {
    for (final c in [_basic, _allowance, _months, _bonuses, _withheld]) {
      c.dispose();
    }
    super.dispose();
  }

  /// Typed text to a number the engine can use: commas and spaces out, blank
  /// and junk to zero. The engine clamps and rounds from there, so this stays
  /// a reader rather than a second place where money rules live.
  num _num(TextEditingController c) {
    final raw = c.text.replaceAll(RegExp(r'[, ]'), '').trim();
    if (raw.isEmpty) return 0;
    final v = num.tryParse(raw);
    return (v == null || !v.isFinite || v < 0) ? 0 : v;
  }

  @override
  Widget build(BuildContext context) {
    final basic = _num(_basic);
    final monthsRaw = _months.text.trim();
    final result = annualizeCompensation(
      basic,
      taxableAllowance: _num(_allowance),
      monthsWorked: monthsRaw.isEmpty ? null : _num(_months),
      bonuses: _num(_bonuses),
      taxWithheld: _num(_withheld),
    );
    final showResult = basic > 0;
    final difference = result['difference'] as double;
    final isRefund = result['isRefund'] as bool;

    return Scaffold(
      appBar: AppBar(
        backgroundColor: Barako.background,
        foregroundColor: Barako.text,
        title: Text(
          'Year-end tax check',
          style: TextStyle(color: Barako.text, fontWeight: FontWeight.w800),
        ),
      ),
      body: SafeArea(
        child: ListView(
          padding: const EdgeInsets.fromLTRB(20, 8, 20, 32),
          children: [
            Text(
              'Every December your employer works out the whole year at once. '
              'Steady monthly tax rarely matches that once a 13th month, a '
              'bonus, a raise, or a mid-year start are counted, so most '
              'people are owed a little back or owe a little more.',
              style: TextStyle(
                color: Barako.textSecondary,
                fontSize: 14,
                height: 1.5,
              ),
            ),
            const SizedBox(height: Gap.lg),
            _label('Monthly basic pay'),
            _field(_basic, hint: 'e.g. 25000'),
            _label('Monthly taxable allowance'),
            _field(_allowance, hint: 'Leave empty if none'),
            _label('Months worked this year'),
            _field(_months, hint: '12 if you worked the whole year'),
            _label('Bonuses and 13th month, for the year'),
            _field(_bonuses, hint: 'Total for the year'),
            _label('Tax already withheld, for the year'),
            _field(_withheld, hint: 'From your payslips'),
            const SizedBox(height: Gap.xl),
            if (!showResult)
              Text(
                'Enter your monthly basic pay to see the estimate.',
                style: TextStyle(color: Barako.muted, fontSize: 13),
              )
            else ...[
              Container(
                padding: const EdgeInsets.all(18),
                decoration: BoxDecoration(
                  color: Barako.card,
                  borderRadius: BorderRadius.circular(Radii.lg),
                  border: Border.all(
                    color: isRefund ? Barako.primary : Barako.warningStrong,
                  ),
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Kicker(isRefund ? 'LIKELY REFUND' : 'LIKELY STILL OWED'),
                    const SizedBox(height: 6),
                    Text(
                      formatMoneyText(difference.abs()),
                      style: TextStyle(
                        fontFamily: Barako.displayFont,
                        color: isRefund
                            ? Barako.primaryText
                            : Barako.warningStrong,
                        fontSize: 30,
                        fontWeight: FontWeight.w700,
                      ),
                    ),
                    const SizedBox(height: 6),
                    Text(
                      isRefund
                          ? 'You paid more through the year than the annual '
                                'computation asks for, so this should come '
                                'back to you.'
                          : 'The annual computation asks for more than was '
                                'withheld, so expect this to be deducted.',
                      style: TextStyle(
                        color: Barako.textSecondary,
                        fontSize: 13,
                        height: 1.4,
                      ),
                    ),
                  ],
                ),
              ),
              const SizedBox(height: Gap.lg),
              Kicker('HOW THAT ADDS UP'),
              const SizedBox(height: 8),
              _row('Regular pay, taxable', result['regularTaxable'] as double),
              _row(
                'Bonuses above the ${formatMoneyText(bonusTaxFreeCeiling)} '
                'tax free ceiling',
                result['bonusTaxable'] as double,
              ),
              _row('Taxable for the year', result['annualTaxable'] as double),
              _row('Annual tax due', result['annualTaxDue'] as double),
              _row('Tax already withheld', result['taxWithheld'] as double),
              const SizedBox(height: Gap.lg),
              Text(
                'An estimate using the ${result['ratesYear']} tables and '
                '${result['monthsWorked']} months of work. It is not your '
                'Form 2316: your employer counts things this cannot see, like '
                'a previous employer that year or benefits treated '
                'differently. Use it to know roughly what is coming, then '
                'check the real thing when it arrives.',
                style: TextStyle(
                  color: Barako.muted,
                  fontSize: 12,
                  height: 1.5,
                ),
              ),
            ],
          ],
        ),
      ),
    );
  }

  Widget _row(String label, double value) => Padding(
    padding: const EdgeInsets.only(bottom: 8),
    child: Row(
      children: [
        Expanded(
          child: Text(
            label,
            style: TextStyle(color: Barako.textSecondary, fontSize: 13),
          ),
        ),
        const SizedBox(width: 12),
        Text(
          formatMoneyText(value),
          style: TextStyle(
            color: Barako.text,
            fontSize: 14,
            fontWeight: FontWeight.w700,
          ),
        ),
      ],
    ),
  );

  Widget _label(String t) => Padding(
    padding: const EdgeInsets.only(top: 14, bottom: 6),
    child: Text(t, style: TextStyle(color: Barako.muted, fontSize: 12)),
  );

  Widget _field(TextEditingController c, {required String hint}) => TextField(
    controller: c,
    keyboardType: const TextInputType.numberWithOptions(decimal: true),
    onChanged: (_) => setState(() {}),
    style: TextStyle(color: Barako.text, fontSize: 16),
    decoration: InputDecoration(
      hintText: hint,
      hintStyle: TextStyle(color: Barako.faint),
      filled: true,
      fillColor: Barako.card,
      border: OutlineInputBorder(
        borderRadius: BorderRadius.circular(Radii.md),
        borderSide: BorderSide(color: Barako.border),
      ),
      enabledBorder: OutlineInputBorder(
        borderRadius: BorderRadius.circular(Radii.md),
        borderSide: BorderSide(color: Barako.border),
      ),
    ),
  );
}
